clash-lib/src/Clash/Normalize/Transformations/Inline.hs

```diff
@@ -393,11 +393,18 @@ in HDL, but not in Haskell, by `unsafeCoerce`.
 The end result of all of this is that we get no/fewer assignments in HDL where the RHS is
 simply a variable reference. See issue #779 -}
 
--- | Takes a binding and collapses its term if it is a noop
+-- | Takes a binding and collapses its term if it is a noop. Only runs at
+-- synthesis boundaries (NOINLINE/OPAQUE functions) to avoid running too early
+-- on functions that might be inlined later. See #3036.
 collapseRHSNoops :: HasCallStack => NormRewrite
-collapseRHSNoops _ (Letrec binds body) = do
-  binds1 <- mapM runCollapseNoop binds
-  return $ Letrec binds1 body
+collapseRHSNoops _ letrec@(Letrec binds body) = do
+  (curFunId, _) <- Lens.use curFun
+  curBinding <- lookupVarEnv curFunId <$> Lens.use bindings
+  case curBinding of
+    Just binding | isOpaque (bindingSpec binding) -> do
```

Comment:
I think `isNoInline` is preferred https://github.com/clash-lang/clash-compiler/blob/a6f3230a26ccff44b076f10c0c74cfe396cc36e8/clash-lib/src/GHC/BasicTypes/Extra.hs#L52-L58

because that is what we use to determine how much of the hierarchy we flatten: https://github.com/clash-lang/clash-compiler/blob/a6f3230a26ccff44b076f10c0c74cfe396cc36e8/clash-lib/src/Clash/Normalize.hs#L298

Reply:
That makes sense